### DISCUSSION TITLE
Custom string on the GUI

### DISCUSSION THREAD

**Discussion by cybermaus** [arendst/Tasmota#15556] [Ideas]
Recently I discovered you can improve both GUI and logic by adding virtual relays to unconnected GPIO pins.
This puts an extra button on the screen, and allows the less technical members of the family some better control without teaching them the console.

Example: In order to make a porchlight connected to a motion detector, I did the following:

- added a Virtual Relay 2 to display&remember whether or not the schedule timers are between sundown and sunup.
- rerouted Input 1 to Virtual Relay 3 so display motion detect 
- added a Virtual PWM 4 dimmer to set the desired delay off time

See screen print.
<img width="509" alt="Capture" src="https://user-images.githubusercontent.com/3198100/167015278-df74c7e7-84e7-42e0-b939-0c88a783eeff.PNG">

Next some local logic, and now the casual user can modify the delay time, as well as override the time window or simply turn on the light manually. All with local logic

While doing this, I realized it could be nice if we could have some option to write a custom string to the GUI.
That would look similar to how energy values are shown, or to how `Webtime` is shown

Right now, I am using/abusing the WebButton4 for this feedback text. 
But I foresee I would also like to display other values.

PS: rules to match above example 
```
Rule1 
  on Power2#State==1 do mem2 1 endon
  on Power2#State==0 do Backlog POWER1 0; mem2 0 endon
  on Power3#State==1 do Backlog RuleTimer1 0; POWER1 %mem2%; endon
  on Power3#State==0 do RuleTimer1 %mem4% endon
  on rules#timer=1 do POWER1 0 endon
  on Dimmer#State do Backlog var4 %value%; mult4 18; endon
  on Var4#State do mem4 %value% endon
  on Mem4#State do WebButton4 %value% seconds endon
Rule1 1
```

Ref: more details on [google group](https://groups.google.com/g/sonoffusers/c/lcMQyksVSrw)

**Comment by barbudor** [comment]
Cool 👍

**Comment by pkkrusty** [comment]
Nice hack of available resources. Will keep these ideas in mind for future projects.

**Comment by barbudor** [comment]
One drawback is that WebButton write the new text to flash. So you may want to use Savedata 0 to reduce wearing.

**Comment by cybermaus** [comment]
Yes, I am doing that. From my notes (I have notes for each device)

```
####################
## Saving settings

# SetOption0 0;   					: No power state saving (flash wear reduce)
# PowerOnState 0;					: Restart as OFF after power loss
# SaveData 30;					    : Reduce flash wear, 
#
# Flashwear strategy is to *not* save power state, as it would flash each individual radar detection pulse.
# However, by moving specifically Timer (Power2) and Delay (PWM4) to %mem% positions, those are saved.
# We only save every 30 seconds though, so not every single PWM slider move is recorded.
# All in all it should record only the timer state twice a day, assuming 100.000 cycles good for 100+ years
#
Backlog SetOption0 0; PowerOnState 0; restart 1
Backlog SaveData 30; Restart 1;
```

But I did not want to write all details in one go. It is all in the mentioned google group though.

**Comment by cybermaus** [comment]
Also, regarding the WebButton remark:

yes, that is of course one of the reasons why I am requesting a feature for custom text. 
The WebButton is a small abuse of the system. That is why this is a idea/feature request.


I am imagining something like similar to `Webtime`. For example:
```
Webtext<x> <text> <value> <unit> <format>
  <x>  maybe up to 4 texts?
  <text>  left afjusted
  <value> right adjusted, but left of <unit>
  <unit> right adjusted
  <format>  0=as is, 1=display time in minutes in hh:mm
```

So in my example, rather then using WebButton, I would use:
``` 
Webtext1 Delay %mem4% seconds 0
```

and it would look something like this.
It would also auto-refresh automatically (I assume, because all such energy values strings are also auto-updates
And avoid WebButton write to flash,

<img width="509" alt="Capture3" src="https://user-images.githubusercontent.com/3198100/167023200-cd2402a1-024c-45aa-be23-2c54beb7d643.PNG">

**Reply by barbudor** [reply]
I never used it myself but Scripting Language has some ability to  customize the web page

**Reply by cybermaus** [reply]
Not used it either, mostly because that requires custom builds and stuff. And memory problems

I do actually build my own, but after initially changing everything, nowadays I try to stay vanilla, in fact all I change is the SSID, so that after a full reset, they come back on my network.

I suspect this is not even a very big change, as all the string routines and templates are already there for the energy/time/temperature sensors. Though ultimately that is for the developers to judge.
